libs/knowledge-store/ragstack_knowledge_store/graph_store.py

```diff
@@ -892,22 +841,27 @@ def _extract_where_clause_params(
 
         return params
 
-    def _get_search_cql(
+    def _get_search_cql_and_params(
         self,
-        has_limit: bool = False,
-        columns: str | None = CONTENT_COLUMNS,
-        metadata_keys: Sequence[str] = (),
-        has_id: bool = False,
-        has_embedding: bool = False,
-        has_link_from_tags: bool = False,
-    ) -> PreparedStatement:
-        where_clause = self._extract_where_clause_cql(
-            has_id=has_id,
-            metadata_keys=metadata_keys,
-            has_link_from_tags=has_link_from_tags,
-        )
-        limit_clause = " LIMIT ?" if has_limit else ""
-        order_clause = " ORDER BY text_embedding ANN OF ?" if has_embedding else ""
+        columns: str,
+        limit: int | None = None,
+        metadata: dict[str, Any] | None = None,
+        embedding: list[float] | None = None,
+        outgoing_link: Link | None = None,
+    ) -> tuple[PreparedStatement|SimpleStatement, tuple[Any, ...]]:
+        if outgoing_link is not None:
+            if metadata is None:
+                metadata = {}
+            else:
+                # don't add link search to original metadata dict
+                metadata = metadata.copy()
+                metadata[_metadata_s_link_key(link=outgoing_link)] = _metadata_s_link_value()
```

Comment:
How does this work? It *looks* like it is doing equality. But, there may be multiple outgoing links, and we need to find the nodes with one of those as an incoming link. It seems like this is maybe going the wrong direction, and also likely missing the set-equality. Are we sure this works the same?

Reply:
when we `add_nodes()`, we are storing all the incoming_links as keys in the `metadata_s` index. There is an arbitrary, static value stored with each key, so that it can be stored in the MAP<text,text> type.

Here, when we search, we are finding all the chunks that have a matching outgoing-link key.

We are essentially doing a hybrid query on metadata keys... the values do not matter.

Reply:
Ah. That's the part I missed. Could you add a comment to that effect where we store / query? It also seems like there is then a risk that the generated key collides with the user defined key? Is there a prefix or something we should tell people *not* to use in their metadata?